File: packages/cli/src/server/stream-registry.ts
import { ProxerError } from "#app/lib/error.ts";
import type { RouteMode } from "#app/protocol/frame.ts";
import type { TunnelConnection } from "#app/protocol/tunnel-connection.ts";
import { type TunnelRoute, tunnelRouteKey } from "#app/server/route-target.ts";

export type TunnelBasicAuth = {
  readonly password: string;
  readonly username?: string;
};

export type RegisteredTunnel = {
  readonly route: TunnelRoute;
  readonly connection: TunnelConnection;
  readonly basicAuth?: TunnelBasicAuth;
};

export type TunnelRegistration = RegisteredTunnel & {
  readonly mode?: RouteMode;
};

export type RegisterTunnelResult = {
  readonly route: TunnelRoute;
  readonly mode: RouteMode;
  readonly replicas: number;
};

type RouteEntry = {
  readonly route: TunnelRoute;
  readonly mode: RouteMode;
  readonly basicAuth?: TunnelBasicAuth;
  readonly tunnels: RegisteredTunnel[];
  nextIndex: number;
};

export class DuplicateTunnelRouteError extends ProxerError {
  readonly route: TunnelRoute;

  constructor(route: TunnelRoute, message = duplicateRouteMessage(route)) {
    super(message);
    this.name = "DuplicateTunnelRouteError";
    this.route = route;
  }
}

export class TunnelRegistry {
  readonly #routes = new Map<string, RouteEntry>();

  register(tunnel: TunnelRegistration): RegisterTunnelResult {
    const key = tunnelRouteKey(tunnel.route);
    const mode = tunnel.mode ?? "single";
    const existing = this.#routes.get(key);
    const registeredTunnel: RegisteredTunnel = {
      ...(tunnel.basicAuth ? { basicAuth: tunnel.basicAuth } : {}),
      connection: tunnel.connection,
      route: tunnel.route,
    };

    if (!existing) {
      this.#routes.set(key, {
        ...(tunnel.basicAuth ? { basicAuth: tunnel.basicAuth } : {}),
        mode,
        nextIndex: 0,
        route: tunnel.route,
        tunnels: [registeredTunnel],
      });
      return { route: tunnel.route, mode, replicas: 1 };
    }

    if (existing.mode !== mode) {
      throw new DuplicateTunnelRouteError(
        tunnel.route,
        duplicateRouteMessage(tunnel.route, existing.mode),
      );
    }

    if (mode === "single") {
      throw new DuplicateTunnelRouteError(tunnel.route);
    }

    if (!sameBasicAuth(existing.basicAuth, tunnel.basicAuth)) {
      throw new ProxerError(
        "Cluster tunnel basic auth must match existing route",
      );
    }

    existing.tunnels.push(registeredTunnel);
    return {
      route: existing.route,
      mode: existing.mode,
      replicas: existing.tunnels.length,
    };
  }

  unregister(route: TunnelRoute, connection?: TunnelConnection): void {
    const key = tunnelRouteKey(route);
    const entry = this.#routes.get(key);
    if (!entry) {
      return;
    }

    if (!connection) {
      this.#routes.delete(key);
      return;
    }

    const tunnelIndex = entry.tunnels.findIndex(
      (tunnel) => tunnel.connection === connection,
    );
    if (tunnelIndex === -1) {
      return;
    }

    entry.tunnels.splice(tunnelIndex, 1);
    if (entry.tunnels.length === 0) {
      this.#routes.delete(key);
      return;
    }

    entry.nextIndex %= entry.tunnels.length;
  }

  get(route: TunnelRoute): RegisteredTunnel | undefined {
    const entry = this.#routes.get(tunnelRouteKey(route));
    if (!entry) {
      return undefined;
    }

    if (entry.mode === "single") {
      return entry.tunnels[0];
    }

    const tunnel = entry.tunnels[entry.nextIndex % entry.tunnels.length];
    entry.nextIndex = (entry.nextIndex + 1) % entry.tunnels.length;
    return tunnel;
  }
}

const duplicateRouteMessage = (
  route: TunnelRoute,
  mode?: RouteMode,
): string => {
  const modeSuffix = mode === undefined ? "" : ` in ${mode} mode`;
  if (route.type === "root") {
    return `Tunnel root domain is already registered${modeSuffix}`;
  }

  return `Tunnel subdomain "${route.subdomain}" is already registered${modeSuffix}`;
};

const sameBasicAuth = (
  left: TunnelBasicAuth | undefined,
  right: TunnelBasicAuth | undefined,
): boolean => {
  return (
    left?.password === right?.password && left?.username === right?.username
  );
};

File: packages/cli/lib/src/services/server/stream_registry.dart
import 'package:proxer/src/services/protocol/frame.dart';
import 'package:proxer/src/services/protocol/tunnel_connection.dart';
import 'package:proxer/src/services/server/route_target.dart';
import 'package:proxer/src/util/error.dart';

class RegisteredTunnel {
  const RegisteredTunnel({
    required this.route,
    required this.connection,
    this.basicAuth,
  });

  final TunnelRoute route;
  final TunnelConnection connection;
  final BasicAuthConfig? basicAuth;
}

class RegisterTunnelResult {
  const RegisterTunnelResult({
    required this.route,
    required this.mode,
    required this.replicas,
  });

  final TunnelRoute route;
  final RouteMode mode;
  final int replicas;
}

class DuplicateTunnelRouteError extends ProxerError {
  DuplicateTunnelRouteError(this.route, [String? message])
    : super(message ?? _duplicateRouteMessage(route));

  final TunnelRoute route;
}

class _RouteEntry {
  _RouteEntry({
    required this.route,
    required this.mode,
    required this.basicAuth,
    required this.tunnels,
  });

  final TunnelRoute route;
  final RouteMode mode;
  final BasicAuthConfig? basicAuth;
  final List<RegisteredTunnel> tunnels;
  int nextIndex = 0;
}

class TunnelRegistry {
  final Map<String, _RouteEntry> _routes = {};

  RegisterTunnelResult register({
    required TunnelRoute route,
    required TunnelConnection connection,
    RouteMode mode = RouteMode.single,
    BasicAuthConfig? basicAuth,
  }) {
    final existing = _routes[route.key];
    final tunnel = RegisteredTunnel(
      route: route,
      connection: connection,
      basicAuth: basicAuth,
    );
    if (existing == null) {
      _routes[route.key] = _RouteEntry(
        route: route,
        mode: mode,
        basicAuth: basicAuth,
        tunnels: [tunnel],
      );
      return RegisterTunnelResult(route: route, mode: mode, replicas: 1);
    }
    if (existing.mode != mode) {
      throw DuplicateTunnelRouteError(
        route,
        '${_duplicateRouteMessage(route)} in ${existing.mode.value} mode',
      );
    }
    if (mode == RouteMode.single) {
      throw DuplicateTunnelRouteError(route);
    }
    if (!_sameBasicAuth(existing.basicAuth, basicAuth)) {
      throw ProxerError('Cluster tunnel basic auth must match existing route');
    }
    existing.tunnels.add(tunnel);
    return RegisterTunnelResult(
      route: existing.route,
      mode: mode,
      replicas: existing.tunnels.length,
    );
  }

  void unregister(TunnelRoute route, [TunnelConnection? connection]) {
    final entry = _routes[route.key];
    if (entry == null) {
      return;
    }
    if (connection == null) {
      _routes.remove(route.key);
      return;
    }
    entry.tunnels.removeWhere(
      (tunnel) => identical(tunnel.connection, connection),
    );
    if (entry.tunnels.isEmpty) {
      _routes.remove(route.key);
    } else {
      entry.nextIndex %= entry.tunnels.length;
    }
  }

  RegisteredTunnel? get(TunnelRoute route) {
    final entry = _routes[route.key];
    if (entry == null) {
      return null;
    }
    if (entry.mode == RouteMode.single) {
      return entry.tunnels.first;
    }
    final tunnel = entry.tunnels[entry.nextIndex % entry.tunnels.length];
    entry.nextIndex = (entry.nextIndex + 1) % entry.tunnels.length;
    return tunnel;
  }
}

String _duplicateRouteMessage(TunnelRoute route) {
  return switch (route) {
    RootTunnelRoute() => 'Tunnel root domain is already registered',
    SubdomainTunnelRoute(:final subdomain) =>
      'Tunnel subdomain "$subdomain" is already registered',
  };
}

bool _sameBasicAuth(BasicAuthConfig? left, BasicAuthConfig? right) {
  return left?.password == right?.password && left?.username == right?.username;
}
